modules/puppet_proxy_mcollective/mcollective_main.rb

```diff
@@ -21,7 +21,8 @@ def run(nodes)
     # Defaults:foreman-proxy !requiretty
     # foreman-proxy ALL=(peadmin) NOPASSWD: /opt/puppet/bin/mco *',
     if user
-      cmd.push("-u", user)
+      homedir = File.expand_path("~#{user}")
```

Comment:
Would it be good to verify ```HOME``` exists and prefer that, but use this as a fallback?

Reply:
I think `HOME` does exist it's just the value contains the home directory for the running user not the impersonated user. I'll do some tests to confirm though.

Reply:
Oh, that's a subtlety I missed. For me it does set the HOME var:
```
$ cat test.py
#!/usr/bin/env python3
import os
print(os.environ['HOME'])
$ ./test.py 
/home/ekohl
$ sudo -u root ./test.py
/root
```

There could be a different interaction in a non-interactive shell though.

Reply:
Ok I think this is actually an ubuntu vs other distros default behaviour thing, sorry about this

```
sysadmin:~$ ruby -e 'puts File.expand_path("~"); puts ENV.fetch("HOME")'
/home/sysadmin
/home/sysadmin
sysadmin:~$ sudo -u root ruby -e 'puts File.expand_path("~"); puts ENV.fetch("HOME")'
/home/sysadmin
/home/sysadmin
```
(same in python)

However setting `Defaults        env_keep -= "HOME"` in `/etc/sudoers` gives the behaviour you are seeing so this PR isn't needed I don't think. Maybe just a note in the docs?

Reply:
I figured it would be something like that. My sudoers (Fedora 27) has `Defaults env_reset` and then doesn't list `HOME` in `env_keep`.

Which version of Ubuntu are you on?
Is this the default config? If so, it'd be nice to implement a solution so it works out of the box.

Reply:
They're stock Ubuntu 16.04.2 LTS with basically just the foreman puppet::server and foreman-proxy puppet modules applied. Perhaps adding another section to the foreman-proxy sudoers.d file (or a separate file) is better?
It would also mean foreman-proxy could run `mco puppet runonce` out of the box too.